consensus/src/main/java/org/apache/iotdb/consensus/ratis/ApplicationStateMachineProxy.java

```diff
@@ -117,23 +119,33 @@ public CompletableFuture<Message> applyTransaction(TransactionContext trx) {
               log.getStateMachineLogEntry().getLogData().asReadOnlyByteBuffer());
     }
 
-    Message ret;
+    Message ret = null;
     waitUntilSystemNotReadOnly();
+    TSStatus finalStatus = null;
+    boolean shouldRetry = false;
     do {
       try {
         TSStatus result = applicationStateMachine.write(applicationRequest);
-        ret = new ResponseMessage(result);
-        break;
+
+        finalStatus =
+            (finalStatus == null) ? result : retryPolicy.updateResult(finalStatus, result);
```

Comment:
use another `boolean` flag to indicate whether this time is first try in case of retryPolicy.updateResult() return a `null` value

Reply:
done